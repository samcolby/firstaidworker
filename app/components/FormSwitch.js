import React, { memo } from "react";
import PropTypes from "prop-types";

import { ListItem } from "react-native-elements";

/**
 * React Pure Component to display
 * a label, switch component and error message
 * @param {Object} props - Standard react props
 * @param {string} fieldName
 *    The name of this field
 * @param {Object} formikProps
 *    The formik props used by this switch
 * @param {string} label
 *    The label for the switch component, defaults to "Field label"
 * @param {Object} switchProps
 *    Extra props to be passed to the switch component
 */
function FormSwitch(props) {
  const { label = "Field label", onValueChange, switchProps, value } = props;

  const toggle = newValue => {
    if (onValueChange && typeof onValueChange === "function") {
      onValueChange(newValue);
    }
    if (switchProps && typeof switchProps.onValueChange === "function") {
      switchProps.onValueChange(newValue);
    }
  };

  return (
    <ListItem
      switch={{
        ...switchProps,
        value: value,
        onValueChange: toggle
      }}
      title={label}
    />
  );
}

FormSwitch.propTypes = {
  fieldName: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onValueChange: PropTypes.func,
  switchProps: PropTypes.object,
  value: PropTypes.bool
};

export default memo(FormSwitch);
