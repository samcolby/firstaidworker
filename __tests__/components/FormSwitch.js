import React from "react";
import { shallow } from "enzyme";

import FormSwitch from "../../app/components/FormSwitch";

const testProps = {
  fieldName: "MyField",
  formikProps: {
    formikValue: 2,
    setFieldValue: jest.fn(),
    values: { MyField: true }
  },
  label: "My field label"
};

describe("<FormSwitch /> component", () => {
  it("renders correctly", () => {
    const component = shallow(<FormSwitch {...testProps} />);

    expect(component.length).toBe(1);
    expect(component).toMatchSnapshot();
  });

  it("displays the correct initialValue", () => {
    const component = shallow(
      <FormSwitch initialValue={true} {...testProps} />
    );
    expect(component.prop("switch").value).toEqual(true);
  });

  it("displays the label", () => {
    const component = shallow(<FormSwitch {...testProps} />);
    expect(component.prop("title")).toBe(testProps.label);
  });

  it("onValueChange on the switch component is called", () => {
    const onValueChange = jest.fn();
    const component = shallow(
      <FormSwitch
        initialValue={false}
        switchProps={{ onValueChange: onValueChange }}
        {...testProps}
      />
    );

    component.prop("switch").onValueChange();

    expect(onValueChange).toHaveBeenCalled();
  });
});
