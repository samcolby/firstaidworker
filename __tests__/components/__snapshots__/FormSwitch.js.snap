// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSwitch /> component renders correctly 1`] = `
<Themed.ListItem
  switch={
    Object {
      "onValueChange": [Function],
      "value": undefined,
    }
  }
  title="My field label"
/>
`;
